### ISSUE TITLE
Magnetic Laplacian for Spectral Info

### ISSUE THREAD

**Issue by mcbosch** [networkx/networkx#8560]
Magnetic Graphs are graphs with complex weights, they have a lot of explanations from a physics and some explaniations in terms of matrices from the ML comunity. I want to add a formal introduction from the Magnetic Graphs in a general and structured way in case this propose seems interesting. 

This PR suggests adding `magnetic_laplacian_matrix` and `magnetic_laplacian_spectrum` to `networkx/linalg/`, extending NetworkX's spectral tools to directed graphs via the magnetic Laplacian.

**Background**
The magnetic Laplacian $L^{(q)}$ is a complex Hermitian matrix associated to a directed graph $G=(V,E)$. Given a parameter $q\in [0, 1/2]$, it is defined as:

$$ L^{(q)} = D - A^{(q)}$$

where D is the degree of  the underlying undirected graph, and $A^{(q)}$ is the Hermitian Magnetic Adjacency Matrix defined as 

$$A^{(q)}_{jk} = e^{2\pi iq\theta_{jk}}\cdot \mathbf{1}_{(j,k)\in \tilde{E}}$$

where $\tilde{E}$ is the set of edges adding the other directions and $\theta_{jk}$ is either 1, -1 or 0 if the edge $(i,j)$ is in the specified direction, oposite or in both in $E$ respectively. Then, $L^{(q)}$ has a spectral descomposition.

This construction appears in Shubin (1994) in a differential-geometric context and was later applied to graph analysis by Fanuel et al. (2017) and to graph neural networks by Zhang et al. (2021, MagNet). The parameter q controls how much directional information is encoded: larger q amplifies asymmetry in the spectrum.

**Motivation**
NetworkX doesn't provide any tool to work with this kind of graphs, and they are very useful. I will also like to scale this PR to classes of functions over the nodes (signals) and over the edges and relate them with differentiation operators. If this issue seems interesting I will add a formal and structured introduction of Magnetic Graph and basic functions.

**Comment by dschult** [comment]
There are so many versions/generalizations of the laplacian Matrix. :)

I think the main question here is whether this is one that is established in the literature, useful and that we can support easily.  The Magentic Laplacian is one of the listed generalizations on the [Wikipeida webpage for Laplacian Matrix](https://en.wikipedia.org/wiki/Laplacian_matrix). And the fairly recent cross from physics literature to ML papers suggests that it is established and probably useful. I suspect it is more useful and popular in the graph neural networks area of graph theory which is not a focus of NetworkX. But it seems reasonable to support a functions for the matrix and another for the spectrum. But perhaps we can show how to compute the spectrum in a docstring example of the function that creates the matrix. That docstring might also be a good place to introduce the concept of magnetic laplacian.

I'm not sure I understand what you mean about eventually adding classes of functions of the nodes and over the edges -- nor how that relates to differential operators.  But a starting PR with the `magnetic_laplacian_matrix` function seems reasonable.

**Comment by mcbosch** [comment]
Thanks for the feedback! 

I saw the Wikipedia webpage for Laplacian Matrix, but differs in some little aspects from more formal approaches. I started the PR with the `magnetic_laplacian_matrix`, read the contributor guide I did it in your format. I hope it can be a good start.

I understand that it's more useful for graph neural network area, I will try to focus more in `networkx` purposes for future contributions or updates. I have already did the PR for the `magnetic_laplacian_matrix`, hope it's all correct. 


By _"adding classes of functions over the nodes and over the edges"_ I meant to express these two Hilbert Spaces:

$$
\ell^2(V) =set( f: V \rightarrow \mathbb{C} |  \sum_{v\in V}|f(v)|^2w_v < \infty),
$$

$$
\ell^2(E) = set(\eta: V \rightarrow \mathbb{C}\;|\; \sum_{e\in E}|\eta(v)|^2w_e < \infty).
$$

Which are related by the exterior derivate $d:\ell^2(V)\rightarrow \ell^2(E)$

$$
df(e) = e^{-i\alpha(e)/2}f(\partial^+e) - e^{i\alpha(e)/2}f(\partial^-e),
$$

where $\partial^+e$ is the pointing node and $\partial^-e$ origin node, and $\alpha$ some potential function. Then it can be also built the adjoint operator $d^*$, so the Magnetic Laplacian is $d^*d$ . This is more used in mathematics, but I thought It would be interesting in graph signal processing. I understand that this is not in the scope of `networkx`, I saw that there is already a package for graph signal processing. So **I will limit my self to the `magnetic_laplacian_matrix` function.** 

N.B: I am trying to find my path in open source and trying to find communities where I can contribute and learn. I really appreciate your response.

**Comment by Whatsonyourmind** [comment]
The magnetic Laplacian is a well-motivated addition. A few mathematical considerations for the implementation:

**Normalization**: The unnormalized version `L^(q) = D - A^(q)` you describe is standard, but many GNN papers (Zhang et al., 2021) use the symmetric normalized form `D^{-1/2} L^(q) D^{-1/2}`. Consider exposing a \`normalized\` parameter, similar to \`nx.normalized_laplacian_matrix\`.

**Spectral properties**: Since `L^(q)` is Hermitian, its eigenvalues are real but `A^(q)` is complex. This means \`scipy.linalg.eigh\` is the correct decomposition (not \`eigvalsh\` on a real matrix). Worth noting in the docstring that the returned spectrum is real-valued despite the complex matrix.

**Edge cases to handle**:
- `q = 0` should recover the standard undirected Laplacian exactly (good test case)
- `q = 0.25` gives maximum directional sensitivity -- useful to mention in docs
- Undirected graphs: every edge `(i,j)` has `theta_ij = 0`, so `L^(q)` reduces to the standard Laplacian for all `q`. This is a useful invariant for testing.

**Performance**: For sparse graphs, the matrix should be returned as \`scipy.sparse\` (CSR format). The phase factors `exp(2*pi*i*q*theta)` are just cube roots of unity {1, exp(2*pi*i*q), exp(-2*pi*i*q)}, so you only need to compute two complex exponentials total, not one per edge.

What convention are you using for bidirectional edges -- treating them as `theta_ij = 0` (symmetric)?

**Comment by mcbosch** [comment]
I am going to handle the aspects you mention to improve performance and use the correct `scipy` function for spectral info. You mention to add a normalized parameter, shouldn't be a new function? I am defining `nx.magnetic_laplacian_matrix` and `nx.normalized_magnetic_laplacian_matrix`.

Regard the conventions, we have to handle some edge cases that maybe would be great to agree in the community.  We have to know how to treat a one directed edge, a bidirected edge, and an undirected edge. Now I am treating bidirectional edges with `theta_ij = 0` and averaging the original weights.  We do this because of how is defined the transformation of from a directed (not symmetric graph) to a symmetric graph (directed):

> Mathematically, the `magnetic_laplacian_matrix` is only defined for symmetric graphs with a potential defined over it. A potential is a function $\alpha: E\rightarrow [0, 2\pi)$, such as if $\overline{e}$ is the opposite direction of $e$, then $\alpha(e) = -\alpha(\overline{e})$.  Then we can assign to each edge $(i,j)$ the weight $\tilde{w}_{ij}e^{i\alpha(e)}$ where $\tilde{w}_{ij}=0.5(w_{ij}+w_{ji})$; and build the magnetic laplacian. Note that this matrix will be Hermitic. Then the transformation from a directed graph to a symmetric graph consists of adding symetries, averaging weights, and for a fixed $q$, define the potential $\alpha(e) = 2\pi q \theta_{ij}$, where $theta_{ij}$ stores the directed info (0 if we had both directions, 1 if it was the original direction and -1 if it's a new direction).

Now, **an important convention should be discussed and agreed**. When we work with an undirected graph, we have to be treated as a symmetric graph. Should the weights be divided? **I think that formally it's better to not split weights for an undirected edge** so it coincides with the Laplacian, **split for directed edges** that aren't bidirectional, and average for bidirectional edges. 

Potentially it would be a good idea to add a parameter to personally define a potential. I am not knowledgeable of the possible applications of this, but it can offer some liberty for those physics or mathematics that study this kind of graph.

**Comment by Whatsonyourmind** [comment]
@mcbosch Good progress. Two design recommendations on the questions you raised:

**Function vs. parameter for normalization.** I would lean toward **a parameter on a single function** rather than two separate functions. The precedent in NetworkX is `nx.laplacian_matrix(G)` and `nx.normalized_laplacian_matrix(G)` as separate functions, but for the magnetic case the additional argument `q` (the magnetic charge) already creates a multi-dimensional parameter space, and adding the normalized variant as a third dimension would lead to an awkward function name explosion. The cleaner API is:

```python
nx.magnetic_laplacian_matrix(G, q=0.25, normalized=False, ...)
```

This matches `scipy.sparse.csgraph.laplacian(csgraph, normed=False)` and is the convention most numerical linear algebra libraries have converged on. Users who want the normalized form pass `normalized=True`. If you want to preserve the existing pair convention, you can wrap it: `nx.normalized_magnetic_laplacian_matrix = partial(nx.magnetic_laplacian_matrix, normalized=True)` — but the underlying implementation lives in one place.

**Edge type semantics for the magnetic phase.** The directed/bidirected/undirected question is the substantive one and the literature is genuinely split. Here is the framework I would use to make the convention defensible:

1. **Directed edge u → v**: applies phase `e^(i * 2πq)` to the (u,v) entry and `e^(-i * 2πq)` to the (v,u) entry. This is the standard convention from Lieb (1993) and the GNN literature (Zhang et al., 2021).

2. **Undirected edge {u, v}**: applies phase `1` to both entries. The magnetic phase encodes orientation, and an unoriented edge has no orientation to encode. The Hermitian structure is preserved trivially.

3. **Bidirected edge (u → v AND v → u explicitly stored as separate edges)**: this is where conventions diverge. Two reasonable choices:
   - **a) Phases cancel**: treat as if undirected (phase = 1). This loses information about whether the user intended two distinct directed edges or one bidirectional relationship.
   - **b) Phases add**: each direction contributes its own phase, so the (u,v) entry gets `e^(i * 2πq) + e^(-i * 2πq) = 2cos(2πq)`. This preserves the multiplicity but breaks the unit-modulus property of phases.

The least-bad default is (a) treat-as-undirected, with a `bidirected_handling='collapse'` parameter that lets users opt into 'sum' or 'separate' if their application requires it. Document the choice in the docstring with a one-line note pointing to the literature.

**A subtle bug to test for.** When `q=0`, the magnetic Laplacian must reduce exactly to the standard Laplacian. This is a one-line invariant that should be a regression test:

```python
assert np.allclose(
    nx.magnetic_laplacian_matrix(G, q=0).toarray(),
    nx.laplacian_matrix(G).toarray()
)
```

This catches off-by-one errors in the phase formula (using `q` vs `2*q`, sign conventions on the phase) at zero overhead.

**For the normalized variant**, the standard form is `D^(-1/2) L^(q) D^(-1/2)` where D is the standard degree matrix (real-valued, computed from the symmetric adjacency without phases). Note that this is the symmetric normalization — the random-walk normalization `D^(-1) L^(q)` is also used in some papers but produces a non-Hermitian operator, which loses the spectral guarantees. I would only ship the symmetric version unless there is explicit user demand for the random-walk form.

Happy to review the PR once you push the design changes. The community-convention question is exactly the kind of thing that benefits from being decided in the issue thread before the PR rather than in PR review.

**Comment by mcbosch** [comment]
Thanks for all the tips. I will work on these design changes and update the PR. For future Issues I will take your advice and discuss more the Issue before going on the PR and maintain it clean. Thank you.

**Comment by mcbosch** [comment]
@Whatsonyourmind can you cite any reference where they collapse the phases? 

I came out with some mathematical issues if we want to collapse. Suppose we have the following graph: $G = (V, E)$ with $V=\lbrace u,v\rbrace$ and $E = \lbrace(u,v),(v,u)\rbrace$. The degree diagonal is $diag(1,1)$. If we want to collapse for example with $q=0.25$ we obtain the identity matrix as the magnetic Laplacian. But if we collapse with $q=0$ we obtain:

$$
L^{0.5} = 
\begin{pmatrix}
1 & -2  \\
-2 & 1  
\end{pmatrix},
$$

Which is not positive-semidefinite matrix. This is quite strange following the theory of the Magnetic Laplacian's derived from physics. If we change the diagonal taking into account the magnitude of the phases, it will be positive semidefinite. All the functions we'll work correctly, but I don't know if it makes sense.

I don't know if there is any application that requires the sum of the phases, and explains a theoretical background of this convention. But if we want to add the collapsing parameter we should think of other conventions, as what happens to the diagonal, we want to normalize the phase $e^{i2\pi q} + e^{-i2\pi q}$, etc. 

I think that if we don't have any theoretical background maybe it's better to restrict ourselves to the default case and in the future add it. It's easy to add. 

Tell me what you think.

**Comment by Whatsonyourmind** [comment]
@mcbosch You are right, and your counter-example is clean — I owe you a correction.

**Your example is the correct failure mode, not an edge case.** Let me walk through why, and then why I should not have suggested the collapse parameter in the first place.

For the bidirectional graph $G = (\{u,v\}, \{(u,v),(v,u)\})$ with the count-degree $D = \operatorname{diag}(1,1)$:

- At $q = 0.25$: off-diagonal is $e^{i \pi/2} + e^{-i \pi/2} = i - i = 0$, so $A = 0$ and $L = D - A = I$. Positive-semidefinite, but the result is degenerate (two fully-disconnected vertices, when in fact they are bidirectionally connected). That is already a signal something is wrong.
- At $q = 0$: off-diagonal is $1 + 1 = 2$, so $A_{uv} = 2$, giving $L = \begin{pmatrix} 1 & -2 \\ -2 & 1 \end{pmatrix}$ with eigenvalues $\{-1, 3\}$. Not PSD.

The root cause is that the "collapse" convention requires the degree matrix to track the **sum of collapsed-edge moduli**, not the count of distinct vertex pairs. With the corrected degree $D_{ii} = \sum_j |A_{ij}(q)|$:
- At $q = 0$: $|A_{uv}| = 2$, so $D = \operatorname{diag}(2,2)$ and $L = \begin{pmatrix} 2 & -2 \\ -2 & 2 \end{pmatrix}$, eigenvalues $\{0, 4\}$ — PSD.
- At $q = 0.25$: $|A_{uv}| = 0$, so $D = 0$ and $L = 0$ — still PSD, but the result is now "no structure" at exactly the $q$ values where the phases cancel, which is a different kind of degeneracy (the null space of $L$ grows whenever the modular arithmetic lines up).

Both of those cases — the $q=0.25$ disconnection and the $q=0$ PSD violation — are structural symptoms that **parallel-edge collapse at matrix-build time is the wrong place to handle directed-graph data**. The standard references (Fanuel–Barthélemy–Courtain–Salva 2018, Furutani–Shibahara–Kitagawa 2020, Zhang–He–Tsochantaridis 2021) all define the magnetic Laplacian on a **symmetrized** adjacency with a separately-tracked phase matrix, exactly to avoid the sum-of-phases-over-parallel-edges situation. The symmetrization happens at graph-construction time (average the weights, record the asymmetry as the phase), and the Laplacian formula never has to decide how to collapse a parallel pair.

**So my earlier advice to add a collapse parameter was premature optimization on my part, and your instinct to restrict the first version to the default case is the right call.** A clean v1 is:

```python
nx.magnetic_laplacian_matrix(G, q, normalized=False, dtype=complex)
```

where $G$ is either a symmetric `Graph` (trivial phase, standard Laplacian) or a `DiGraph` that gets symmetrized internally using the standard construction you already described ($\tilde{w}_{ij} = 0.5(w_{ij} + w_{ji})$, phase from $\theta_{ij} \in \{-1, 0, 1\}$). No collapse parameter, no parallel-edge handling in the Laplacian code path. If someone later needs a non-standard convention, they can pass a pre-symmetrized graph in.

If you want to expose more liberty for physics/GSP users later, the cleanest extension point is a **custom phase function** `potential=callable(u,v,data)` rather than a collapse flag — it gives the user full control over $\alpha(e)$ without any risk of PSD violation inside NetworkX code. But even that is v2; v1 should ship with just the default.

Your call to keep v1 conservative is the right one, and it matches how every reference implementation I know of handles this. Thanks for actually running the counterexample instead of taking my word for it — that is the part that matters.

**Comment by dean985** [comment]
@dschult the thread has since surfaced some design questions (bidirectional edge conventions, normalized parameter vs separate function) that could use your input.. Would be great to hear your take.

  @mcbosch The counterexample you found on phase-collapse breaking positive-semidefiniteness was really nice. 
Regarding the normalized parameter vs separate function question, I think it's worth noting that NetworkX already uses separate functions for laplacian_matrix / normalized_laplacian_matrix, so following that pattern might be more consistent than adding a boolean flag. But that's probably a question for the maintainers.

  @Whatsonyourmind A couple of things I wanted to flag:

  - You mention the phase factors are "cube roots of unity"- I don't think that's right for general q ∈ [0, 0.5]. They're exp(2πiq) which are only roots of unity for specific rational values of q. You can precompute three distinct values
  since θ ∈ {-1, 0, 1}, but the "cube roots of unity" framing is misleading.
  - The bidirected_handling='collapse' suggestion ended up being shown to break PSD guarantees by mcbosch's example. I think that's a good argument for keeping v1 simple and only adding parameters like that if there's a concrete use case
  backed by literature. Adding knobs that can silently produce invalid matrices seems risky.
  - Minor thing-- recommending scipy conventions over NetworkX's own existing API patterns for a function that lives in NetworkX felt a bit off to me. Consistency within the library seems more important.

**Comment by rossbar** [comment]
@dean985 @Whatsonyourmind please refrain from posting AI-generated comments. This is a violation of NetworkX's [automated contribution policy](https://networkx.org/documentation/latest/developer/contribute.html#automated-contributions-policy) and detracts from the discussion.

**Comment by Whatsonyourmind** [comment]
@rossbar You are right to flag this. I apologize for the noise — I did not check the automated contribution policy before engaging and that is on me. I will stop posting on networkx from now on. Thank you for the specific link and for holding the standard consistently.

**Comment by mcbosch** [comment]
This discussion was aimed to simply add a `nx.magnetic_laplacian_matrix`  function and one for the spectral info. I've done it like that in the PR. So far, this thread has been more disrupting than helpful. I think it's better to stop for a while this issue, if a maintainer wants to merge the PR with the functions will do it. 

The ML community has its own functions of the magnetic Laplacian in PyTorch to work with GNNs, so for now there's no urge in adding this Laplacian. If in the future the physic or GSP community feels the need of these functions can reopen this or ask a maintainer to add the PR. If they also require the freedom of defining their own potential for the Laplacian they can contribute to the PR by adding a parameter `potential` with a dict or callable.

**Comment by rossbar** [comment]
I agree @mcbosch , the discussion here is a bit meandering. I think the main purpose (i.e. proposing the addition of magnetic Laplacian) has been fulfilled, so let's use the PR for any continuing discussion.
